### PR TITLE
feat: Bulk issue creation — multi-proposal batches with single approval

### LINKED ISSUE #122: feat: Bulk issue creation — multi-proposal batches with single approval
## Problem

When BM identifies multiple issues to file (e.g. from a codebase audit), it currently proposes them **one at a time**. Each requires a separate ✅ reaction before moving to the next. Filing 8 issues from a single audit meant 8 ✅ reactions plus 8 round-trips through the agent loop — tedious and slow.

Worse: the bot recently told the user *"Next time you say 'confirm all' or 'create all remaining', I'll fire all the create_issue calls in parallel in one shot"* — but **no code path actually implements that promise**. It was saved as a KB entry, which biases future prompts but does not create a capability. This is a false-capability bug.

## Root causes

1. **Shape bottleneck** in `src/lib/claude.ts` — `AgentResult.issueProposal` is singular. `executeToolsInParallel` overwrites it on each `create_issue` tool-use, so only the last survives even though the agent loop already dispatches multiple tool calls in parallel per round.
2. **Brittle confirmation** — `src/app/api/slack/route.ts` re-parses proposals from the bot's own message text via `parseProposalFromMessage`. Fine for one issue; N proposals round-tripped through Slack formatting is a recipe for drift.
3. **No text-command path** — the reaction handler is the only approval channel. \`"confirm all"\` in the thread does nothing.

## Fix

- Change \`AgentResult\` to \`issueProposals: IssueProposal[]\` (always an array; empty means no proposal).
- Add \`src/lib/issue-batch.ts\` with pure helpers: \`formatBatchProposalMessage\`, \`isBulkConfirmText\`, \`summarizeBatchResult\`.
- Persist pending batches in KV under \`pending-issue-batch:{channel}:{messageTs}\` (primary) and \`pending-issue-batch:thread:{channel}:{threadTs}\` (thread pointer). 24h TTL. Replaces message-text round-tripping.
- Two approval paths converge on a single \`executeBatchCreation\` helper:
  1. ✅ reaction on the proposal message
  2. Short thread text like \`"confirm all"\`, \`"yes"\`, \`"create all"\`, \`"go ahead"\`, \`"approve all"\` — thread-scoped (any thread participant)
- Parallel GitHub creation via \`Promise.allSettled\` — per-issue failures do not abort the rest.
- Nudge in the system prompt: multiple \`create_issue\` calls in one turn are supported and will be batch-confirmed.

## Observability (new events, mirror \`kv_op\` schema)

- \`issue_batch_proposed\` — \`{count, threadTs, sampleTitles, requestingUser}\`
- \`issue_batch_confirmed\` — \`{count, confirmVia: \"reaction\"|\"text\", confirmingUser, latencyMs}\`
- \`issue_batch_created\` — \`{totalCount, successCount, failureCount, durationMs, numbers}\`
- \`issue_create_error\` — \`{title, errorClass, errorMessage}\` + \`Sentry.captureException\` tagged \`{flow: \"issue_create\", batchSize}\`

## Acceptance Criteria

- [ ] \`AgentResult.issueProposals\` is \`IssueProposal[]\` (always an array)
- [ ] A single-proposal batch renders the same familiar \"Proposed Issue:\" block as today (no regression)
- [ ] A multi-proposal batch renders a numbered list with titles, labels, and a footer prompting ✅ / \"confirm all\"
- [ ] ✅ reaction on a batch proposal creates every issue in parallel; reply summarizes with per-issue links
- [ ] Thread text \"confirm all\" (and synonyms) on a pending batch triggers the same creation flow
- [ ] \`isBulkConfirmText\` rejects loose matches (e.g. \"yes it failed\", \"create this one\") — only short, intent-only phrases
- [ ] Per-issue failures do not abort the rest; summary lists both created and failed
- [ ] All new observability events emit with the documented fields
- [ ] New unit tests in \`src/lib/issue-batch.test.ts\`; updated test in \`src/lib/claude.test.ts\` verifying multiple \`create_issue\` blocks produce \`issueProposals\` of length N
- [ ] Manual verification in Slack: propose 3, ✅ → all 3 filed; propose 3, say \"confirm all\" → all 3 filed

## Files

- \`src/lib/claude.ts\` — \`AgentResult\` shape change + prompt nudge
- \`src/lib/issue-batch.ts\` — new pure helpers
- \`src/lib/issue-batch.test.ts\` — new unit tests
- \`src/app/api/slack/route.ts\` — batch proposal rendering, KV-backed pending-batch, text-command path, reaction handler
- \`src/tools/create-issue.ts\` — unchanged (legacy parser stays as dead code for 24h then can be removed in a follow-up)
- \`src/lib/claude.test.ts\` — shape-change test
- \`docs/features/issue-creation.md\` (new or updated) — document the batch flow

## Out of scope

- Per-issue picking from the batch (\"create #1 and #3 only\") — adds UX surface for minor gain; file a follow-up if needed
- Slack Block Kit rewrite — stays plain mrkdwn to match current style
- Migration of in-flight legacy single-proposal messages — 24h TTL drains them

### PR DESCRIPTION
**Pull request by vlad-ko**
Closes #122.

## Summary

- Agent can now propose multiple issues in a single turn; approval is one ✅ reaction or a short "confirm all" reply.
- Shape change: `AgentResult.issueProposal` → `issueProposals: IssueProposal[]` (always an array). Single-proposal UX unchanged.
- New pure module `src/lib/issue-batch.ts`: `formatBatchProposalMessage`, `isBulkConfirmText`, `summarizeBatchResult`. 57 unit tests.
- Batches persist in KV under `pending-issue-batch:{channel}:{firstTs}` + thread pointer `pending-issue-batch:thread:{channel}:{threadTs}`. 24 h TTL.
- Concurrency: atomic `kv.del` claim — racing ✅ + "confirm all" can't double-fire.
- Per-issue GitHub failures via `Promise.allSettled` → summary reply lists creates and failures. Per-issue `Sentry.captureException` tagged `{flow: "issue_create", batchSize}`.

## Why

Previous flow required one ✅ per issue. Filing 8 issues from a single audit = 8 reactions + 8 agent round-trips. Worse: the bot recently told the user *"Next time you say 'confirm all' I'll fire them in parallel"* — but no code implemented that promise. It was saved only as a KB entry, which is a false-capability bug. This PR makes the capability real.

## Observability (new events)

| Event | When |
|---|---|
| `issue_batch_proposed` | After posting a batch proposal |
| `issue_batch_confirmed` | Claim succeeded — `confirmVia: "reaction" \| "text"` |
| `issue_batch_claim_lost` | Raced another handler |
| `issue_batch_created` | All creations settled — `{totalCount, successCount, failureCount, durationMs, numbers}` |
| `issue_create_error` | Per-issue failure |

## Test plan

- [x] `npm test` → 500/500 pass
- [x] `npx tsc --noEmit` clean
- [ ] **Manual in Slack sandbox** — single-proposal regression: request one issue, react ✅, verify filed
- [ ] **Manual in Slack sandbox** — batch proposal: request 3 issues, react ✅, verify all 3 filed with summary
- [ ] **Manual in Slack sandbox** — batch proposal: request 3 issues, reply `"confirm all"`, verify all 3 filed
- [ ] **Manual in Slack sandbox** — bulk-confirm strictness: with a pending batch, reply `"yes it failed"` → does NOT trigger creation; falls through to the agent
- [ ] **Manual in Slack sandbox** — race: react ✅ AND type `"confirm all"` near-simultaneously; exactly one creation summary posts; the other handler logs `issue_batch_claim_lost`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Agent can propose multiple GitHub issues in one turn with compact batch formatting.
  * Batch proposals require explicit bulk confirmation (✅ reaction or specific thread reply) to create issues.
  * Confirmations are handled atomically to prevent duplicate issue creation and produce a combined in-thread summary with per-issue results.

* **Documentation**
  * New docs detailing batch proposal/confirmation lifecycle, observability events, and edge cases.

* **Tests**
  * Added tests covering batch formatting, bulk-confirm matching, and batch result summarization.

### CODE REVIEW

File: src/app/api/slack/route.ts [hunk 1]
```diff
@@ -45,6 +51,125 @@ interface PendingCorrection {
...
+    durationMs: Date.now() - createStart,
+    numbers: outcomes
+      .filter((o) => o.status === "success")
+      .map((o) => (o as Extract<BatchCreationOutcome, { status: "success" }>).issue.number),
+    channel,
+    threadTs,
+  });
+
+  await replyInThread(channel, threadTs, summarizeBatchResult(outcomes));
+  return { claimed: true };
```
[the remark this answers is not in the document]
Comment (vlad-ko) [reply src/app/api/slack/route.ts:193]: Pushed [31876fd](https://github.com/vlad-ko/battle-mage/commit/31876fd) that tombstones the claim (Finding 1) and fixes the doc wording (Finding 3).

On this finding (non-retryable failures): I considered the two-phase claim but decided against it. Reasoning:

1. **Partial failures are already handled.** `Promise.allSettled` isolates per-issue failures; the summary lists both created and failed proposals so the user can re-prompt for just the failures.
2. **`createIssue` is not idempotent.** A retry-from-processing-state mechanism would need a dedupe key (e.g. `idempotency-key-{firstTs}-{index}` mapped to "created GitHub #N") before retrying could be safe. That's a bigger protocol than this workflow warrants.
3. **Catastrophic-failure recovery path already exists.** If everything between `kv.del` and `replyInThread` throws, the outer try/catch in the reaction handler posts a fallback error reply. Issues are on GitHub; the user sees an error; they can check GitHub to see what was created. Not pretty, but not lossy.
4. **The user can always re-propose.** The agent has the source material; re-running the same prompt regenerates the same batch. For an interactive Slack workflow, this is more robust than a retry protocol that could create duplicates.

Happy to revisit if we see this actually bite in prod — I'll add a follow-up if `issue_create_error` rates show meaningful losses.
